src/SourceLocator/Ast/Parser/MemoizingParser.php

```diff
@@ -34,15 +38,25 @@ public function parse(string $code, ErrorHandler|null $errorHandler = null): arr
         $hash = sprintf('%s:%d', hash('sha256', $code), strlen($code));
 
         if (array_key_exists($hash, $this->sourceHashToAst)) {
+            [$serializedAst, $tokens] = $this->sourceHashToAst[$hash];
             /** @var Node\Stmt[]|null $ast */
-            $ast = unserialize($this->sourceHashToAst[$hash]);
+            $ast              = unserialize($serializedAst);
+            $this->lastTokens = $tokens;
 
             return $ast;
         }
 
         $ast                          = $this->wrappedParser->parse($code, $errorHandler);
-        $this->sourceHashToAst[$hash] = serialize($ast);
+        $tokens                       = $this->wrappedParser->getTokens();
+        $this->sourceHashToAst[$hash] = [serialize($ast), $tokens];
+        $this->lastTokens             = $tokens;
 
         return $ast;
     }
+
+    /** @return Token[] */
+    public function getTokens(): array
+    {
+        return $this->lastTokens;
```

Comment:
I think we should call `$this->wrappedParser->getTokens()` directly, and not memoize ourselves.

If this API is added to be compliant with the upstream interface, let's just re-use the memoization from upstream, and not redo it ourselves :)

Reply:
After some pondering, I think we can leave this as-is: forwarding to the wrapped parser wouldn't work anyway, when returning cached results.